src/mapping/nav2_map_server/src/map_reps/occupancy_grid.cpp

```diff
@@ -0,0 +1,324 @@
+  } catch (YAML::Exception) {
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+      "The map does not contain a resolution tag or it is invalid.");
+    throw std::runtime_error("The map does not contain a resolution tag or it is invalid.");
+  }
+
+  try {
+    doc["negate"] >> negate;
+  } catch (YAML::Exception) {
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+      "The map does not contain a negate tag or it is invalid.");
+    throw std::runtime_error("The map does not contain a negate tag or it is invalid.");
+  }
+
+  try {
+    doc["occupied_thresh"] >> occ_th;
+  } catch (YAML::Exception) {
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+      "The map does not contain an occupied_thresh tag or it is invalid.");
+    throw std::runtime_error("The map does not contain an occupied_thresh tag or it is invalid.");
+  }
+
+  try {
+    doc["free_thresh"] >> free_th;
+  } catch (YAML::Exception) {
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+      "The map does not contain a free_thresh tag or it is invalid.");
+    throw std::runtime_error("The map does not contain a free_thresh tag or it is invalid.");
+  }
+
+  try {
+    std::string modeS = "";
+    doc["mode"] >> modeS;
+    if (modeS == "trinary") {
+      mode = TRINARY;
+    } else if (modeS == "scale") {
+      mode = SCALE;
+    } else if (modeS == "raw") {
+      mode = RAW;
+    } else {
+      RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+        "Invalid mode tag \"%s\".", modeS.c_str());
+      throw std::runtime_error("Invalid mode tag.");
+    }
+  } catch (YAML::Exception) {
+    RCLCPP_DEBUG(rclcpp::get_logger("map_server"),
+      "The map does not contain a mode tag or it is invalid.");
+    mode = TRINARY;
+  }
+
+  try {
+    doc["origin"][0] >> origin[0];
+    doc["origin"][1] >> origin[1];
+    doc["origin"][2] >> origin[2];
+  } catch (YAML::Exception) {
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+      "The map does not contain an origin tag or it is invalid.");
+    throw std::runtime_error("The map does not contain an origin tag or it is invalid.");
+  }
+
+  try {
+    doc["image"] >> map_name_;
+    // TODO(bpwilcox): make this path-handling more robust
+    if (map_name_.size() == 0) {
+      RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+        "The image tag cannot be an empty string.");
+      throw std::runtime_error("The image tag cannot be an empty string.");
+    }
+    if (map_name_[0] != '/') {
+      // dirname can modify what you pass it
+      char * fname_copy = strdup(file_name.c_str());
+      map_name_ = std::string(dirname(fname_copy)) + '/' + map_name_;
+      free(fname_copy);
+    }
+  } catch (YAML::Exception) {
+    RCLCPP_ERROR(rclcpp::get_logger("map_server"),
+      "The map does not contain an image tag or it is invalid.");
+    throw std::runtime_error("The map does not contain an image tag or it is invalid.");
+  }
+}
+
+void OccGridLoader::LoadMapFromFile(std::string map_name_)
```

Comment:
Isn't this the PGM loading function?  I'd rename to `loadMapFromPGM`

Reply:
This will also load from bmp. Since it is derived from the base interface, I think LoadMapFromFile is generic enough to be applicable to future extensions of map representations.

Reply:
PNGs are handled differently, no?  Perhaps make the loadMapFromFile function check to see what the extension of the file is and load it appropriately.